### PR TITLE
feat(cli): add --debug-port and --wait-for-client options to dev command

### PR DESCRIPTION
**Pull request by mgrunberg**
## Description

This pull request adds support for remote debugging to the `dev` command in the Aegra CLI by introducing new options for running the server under `debugpy`. It also includes updates to the documentation and comprehensive tests for the new debugging features.

## Type of Change

- [x] `feat`: New feature
- [ ] `fix`: Bug fix
- [ ] `docs`: Documentation changes
- [ ] `style`: Code style/formatting
- [ ] `refactor`: Code refactoring
- [ ] `perf`: Performance improvement
- [ ] `test`: Tests added/updated
- [ ] `chore`: Maintenance (dependencies, build, etc.)
- [ ] `ci`: CI/CD changes

## Related Issues

None

## Changes Made

- Added `--debug-port` and `--wait-for-client` options to the `dev` CLI command, allowing the server to be run under `debugpy` for remote debugging and optionally wait for a debugger client to attach before starting. 
- Modified the process startup logic in `cli.py` to wrap the `uvicorn` server with `debugpy` when `--debug-port` is specified, including error handling for missing `debugpy` and conditional addition of the `--wait-for-client` flag.
- Improved CLI output to display debugging-related information when these options are used.
- Updated the CLI reference documentation to describe the new `--debug-port` and `--wait-for-client` options. 

## Testing

- [x] Unit tests added/updated
- [ ] Integration tests added/updated
- [ ] E2E tests added/updated
- [x] Manual testing performed

## Checklist

- [ ] Code follows project style (Ruff formatting)
- [ ] All linting checks pass (`make lint`)
- [ ] Type checking passes (`make type-check`)
- [ ] All tests pass (`make test`)
- [ ] Documentation updated (if needed)
- [ ] Commit messages follow [Conventional Commits](https://www.conventionalcommits.org/)
- [ ] PR title follows Conventional Commits format

## Summary by CodeRabbit

* **New Features**
  * CLI options to run the dev server under a debugger: --debug-port, --debug-host, and --wait-for-client
* **Improvements**
  * Clear error when debugging tooling is missing; warning shown for non-loopback debug hosts; startup output displays debug settings
* **Tests**
  * Added CLI tests for debug-mode behavior, wait-for-client, host exposure warning, and missing-tool error
* **Chores**
  * Package versions bumped to 0.9.3; optional debug dependency declared for debug installs

<h3>Greptile Summary</h3>

Adds `--debug-port` and `--wait-for-client` options to the `dev` command, wrapping uvicorn with `debugpy` when a port is supplied, with a graceful error when `debugpy` is not installed. The implementation is clean and tests cover the main scenarios; two minor polish items remain (UX ordering of the debugpy check and test patch scope).

<h3>Confidence Score: 5/5</h3>

Safe to merge — all remaining findings are P2 style/improvement suggestions with no correctness impact

Prior P1 concerns (silent no-op, unreachable branch, reload advisory) were addressed in previous review rounds. The two remaining comments are non-blocking: info-panel ordering is a UX polish issue and the global find_spec patch is unlikely to cause test failures in practice. Feature logic and tests are correct.

No files require special attention

<details open><summary><h3>Vulnerabilities</h3></summary>

No security concerns identified. The `debugpy` listen address inherits the `--host` value — users should be aware that passing `--host 0.0.0.0 --debug-port <port>` exposes the debug port on all interfaces, but this is consistent with the existing host flag behaviour and not a defect introduced by this PR.
</details>

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| libs/aegra-cli/src/aegra_cli/cli.py | Adds --debug-port/--wait-for-client options; debugpy check happens after info panel is printed (minor UX ordering issue); logic is otherwise correct |
| libs/aegra-cli/tests/test_cli.py | Three new tests cover the happy path, --wait-for-client flag, and missing-debugpy error; global find_spec patch could silently shadow other module lookups during invoke |
| docs/reference/cli.mdx | Two new option rows added to the dev command table; descriptions are accurate and complete |

</details>

<h3>Flowchart</h3>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[aegra dev --debug-port PORT] --> B{debug_port set?}
    B -- No --> C[Build uvicorn cmd]
    B -- Yes --> D{debugpy installed?}
    D -- No --> E[Print error & exit 1]
    D -- Yes --> F[Print reload advisory]
    F --> G{--wait-for-client?}
    G -- Yes --> H["cmd = [python, -m, debugpy, --listen, host:port, --wait-for-client, -m, uvicorn, ...]"]
    G -- No --> I["cmd = [python, -m, debugpy, --listen, host:port, -m, uvicorn, ...]"]
    C --> J["cmd = [python, -m, uvicorn, ...]"]
    H --> K[subprocess.Popen]
    I --> K
    J --> K
    K --> L[Wait for process to exit]
```

<a href="https://app.greptile.com/ide/claude-code?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Alibs%2Faegra-cli%2Fsrc%2Faegra_cli%2Fcli.py%3A342-376%0A**Info%20panel%20rendered%20before%20%60debugpy%60%20availability%20check**%0A%0AThe%20info%20panel%20that%20says%20%22Debug%20port%3A%205678%22%20is%20printed%20at%20line%20~348%2C%20but%20the%20%60importlib.util.find_spec%28%22debugpy%22%29%60%20check%20only%20runs%20at%20line%20~360.%20If%20%60debugpy%60%20is%20not%20installed%20the%20user%20sees%20a%20panel%20announcing%20the%20debug%20port%20is%20configured%2C%20immediately%20followed%20by%20an%20error%20saying%20%60debugpy%60%20is%20not%20installed%20%E2%80%94%20which%20is%20confusing.%0A%0AMoving%20the%20%60find_spec%60%20check%20before%20the%20%60console.print%28Panel%28...%29%29%60%20call%20would%20give%20a%20cleaner%20failure%20path.%0A%0A%60%60%60python%0A%20%20%20%20%23%20Check%20debugpy%20before%20printing%20the%20info%20panel%0A%20%20%20%20if%20debug_port%20is%20not%20None%3A%0A%20%20%20%20%20%20%20%20if%20importlib.util.find_spec%28%22debugpy%22%29%20is%20None%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20console.print%28%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22%5Bbold%20red%5DError%3A%5B%2Fbold%20red%5D%20debugpy%20is%20not%20installed.%5Cn%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22Install%20it%20with%3A%20%5Bcyan%5Dpip%20install%20debugpy%5B%2Fcyan%5D%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%29%0A%20%20%20%20%20%20%20%20%20%20%20%20sys.exit%281%29%0A%0A%20%20%20%20%23%20Build%20info%20panel%20content%0A%20%20%20%20info_lines%20%3D%20%5B...%5D%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Alibs%2Faegra-cli%2Ftests%2Ftest_cli.py%3A268-270%0A**Global%20%60find_spec%60%20patch%20may%20shadow%20non-%60debugpy%60%20lookups**%0A%0A%60patch%28%22importlib.util.find_spec%22%2C%20return_value%3Dobject%28%29%29%60%20replaces%20the%20function%20for%20the%20entire%20duration%20of%20the%20test.%20Any%20call%20to%20%60find_spec%60%20made%20by%20Click%20internals%20or%20other%20code%20paths%20during%20the%20invoke%20will%20also%20return%20%60object%28%29%60%20instead%20of%20the%20real%20result.%20This%20same%20pattern%20is%20used%20in%20the%20other%20debug%20test.%0A%0AScope%20the%20mock%20to%20%60%22debugpy%22%60%20only%20with%20a%20%60side_effect%60%3A%0A%0A%60%60%60python%0Aoriginal_find_spec%20%3D%20importlib.util.find_spec%0A%0Adef%20fake_find_spec%28name%3A%20str%29%20-%3E%20object%20%7C%20None%3A%0A%20%20%20%20if%20name%20%3D%3D%20%22debugpy%22%3A%0A%20%20%20%20%20%20%20%20return%20object%28%29%0A%20%20%20%20return%20original_find_spec%28name%29%0A%0Awith%20patch%28%22importlib.util.find_spec%22%2C%20side_effect%3Dfake_find_spec%29%3A%0A%20%20%20%20...%0A%60%60%60%0A%0A&repo=ibbybuilds%2Faegra">Fix All in Claude Code</a> <a href="https://app.greptile.com/api/ide/cursor?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Alibs%2Faegra-cli%2Fsrc%2Faegra_cli%2Fcli.py%3A342-376%0A**Info%20panel%20rendered%20before%20%60debugpy%60%20availability%20check**%0A%0AThe%20info%20panel%20that%20says%20%22Debug%20port%3A%205678%22%20is%20printed%20at%20line%20~348%2C%20but%20the%20%60importlib.util.find_spec%28%22debugpy%22%29%60%20check%20only%20runs%20at%20line%20~360.%20If%20%60debugpy%60%20is%20not%20installed%20the%20user%20sees%20a%20panel%20announcing%20the%20debug%20port%20is%20configured%2C%20immediately%20followed%20by%20an%20error%20saying%20%60debugpy%60%20is%20not%20installed%20%E2%80%94%20which%20is%20confusing.%0A%0AMoving%20the%20%60find_spec%60%20check%20before%20the%20%60console.print%28Panel%28...%29%29%60%20call%20would%20give%20a%20cleaner%20failure%20path.%0A%0A%60%60%60python%0A%20%20%20%20%23%20Check%20debugpy%20before%20printing%20the%20info%20panel%0A%20%20%20%20if%20debug_port%20is%20not%20None%3A%0A%20%20%20%20%20%20%20%20if%20importlib.util.find_spec%28%22debugpy%22%29%20is%20None%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20console.print%28%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22%5Bbold%20red%5DError%3A%5B%2Fbold%20red%5D%20debugpy%20is%20not%20installed.%5Cn%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22Install%20it%20with%3A%20%5Bcyan%5Dpip%20install%20debugpy%5B%2Fcyan%5D%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%29%0A%20%20%20%20%20%20%20%20%20%20%20%20sys.exit%281%29%0A%0A%20%20%20%20%23%20Build%20info%20panel%20content%0A%20%20%20%20info_lines%20%3D%20%5B...%5D%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Alibs%2Faegra-cli%2Ftests%2Ftest_cli.py%3A268-270%0A**Global%20%60find_spec%60%20patch%20may%20shadow%20non-%60debugpy%60%20lookups**%0A%0A%60patch%28%22importlib.util.find_spec%22%2C%20return_value%3Dobject%28%29%29%60%20replaces%20the%20function%20for%20the%20entire%20duration%20of%20the%20test.%20Any%20call%20to%20%60find_spec%60%20made%20by%20Click%20internals%20or%20other%20code%20paths%20during%20the%20invoke%20will%20also%20return%20%60object%28%29%60%20instead%20of%20the%20real%20result.%20This%20same%20pattern%20is%20used%20in%20the%20other%20debug%20test.%0A%0AScope%20the%20mock%20to%20%60%22debugpy%22%60%20only%20with%20a%20%60side_effect%60%3A%0A%0A%60%60%60python%0Aoriginal_find_spec%20%3D%20importlib.util.find_spec%0A%0Adef%20fake_find_spec%28name%3A%20str%29%20-%3E%20object%20%7C%20None%3A%0A%20%20%20%20if%20name%20%3D%3D%20%22debugpy%22%3A%0A%20%20%20%20%20%20%20%20return%20object%28%29%0A%20%20%20%20return%20original_find_spec%28name%29%0A%0Awith%20patch%28%22importlib.util.find_spec%22%2C%20side_effect%3Dfake_find_spec%29%3A%0A%20%20%20%20...%0A%60%60%60%0A%0A">Fix All in Cursor</a> <a href="https://app.greptile.com/ide/codex?prompt=Fix%20the%20following%202%20code%20review%20issues.%20Work%20through%20them%20one%20at%20a%20time%2C%20proposing%20concise%20fixes.%0A%0A---%0A%0A%23%23%23%20Issue%201%20of%202%0Alibs%2Faegra-cli%2Fsrc%2Faegra_cli%2Fcli.py%3A342-376%0A**Info%20panel%20rendered%20before%20%60debugpy%60%20availability%20check**%0A%0AThe%20info%20panel%20that%20says%20%22Debug%20port%3A%205678%22%20is%20printed%20at%20line%20~348%2C%20but%20the%20%60importlib.util.find_spec%28%22debugpy%22%29%60%20check%20only%20runs%20at%20line%20~360.%20If%20%60debugpy%60%20is%20not%20installed%20the%20user%20sees%20a%20panel%20announcing%20the%20debug%20port%20is%20configured%2C%20immediately%20followed%20by%20an%20error%20saying%20%60debugpy%60%20is%20not%20installed%20%E2%80%94%20which%20is%20confusing.%0A%0AMoving%20the%20%60find_spec%60%20check%20before%20the%20%60console.print%28Panel%28...%29%29%60%20call%20would%20give%20a%20cleaner%20failure%20path.%0A%0A%60%60%60python%0A%20%20%20%20%23%20Check%20debugpy%20before%20printing%20the%20info%20panel%0A%20%20%20%20if%20debug_port%20is%20not%20None%3A%0A%20%20%20%20%20%20%20%20if%20importlib.util.find_spec%28%22debugpy%22%29%20is%20None%3A%0A%20%20%20%20%20%20%20%20%20%20%20%20console.print%28%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22%5Bbold%20red%5DError%3A%5B%2Fbold%20red%5D%20debugpy%20is%20not%20installed.%5Cn%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%22Install%20it%20with%3A%20%5Bcyan%5Dpip%20install%20debugpy%5B%2Fcyan%5D%22%0A%20%20%20%20%20%20%20%20%20%20%20%20%29%0A%20%20%20%20%20%20%20%20%20%20%20%20sys.exit%281%29%0A%0A%20%20%20%20%23%20Build%20info%20panel%20content%0A%20%20%20%20info_lines%20%3D%20%5B...%5D%0A%60%60%60%0A%0A%23%23%23%20Issue%202%20of%202%0Alibs%2Faegra-cli%2Ftests%2Ftest_cli.py%3A268-270%0A**Global%20%60find_spec%60%20patch%20may%20shadow%20non-%60debugpy%60%20lookups**%0A%0A%60patch%28%22importlib.util.find_spec%22%2C%20return_value%3Dobject%28%29%29%60%20replaces%20the%20function%20for%20the%20entire%20duration%20of%20the%20test.%20Any%20call%20to%20%60find_spec%60%20made%20by%20Click%20internals%20or%20other%20code%20paths%20during%20the%20invoke%20will%20also%20return%20%60object%28%29%60%20instead%20of%20the%20real%20result.%20This%20same%20pattern%20is%20used%20in%20the%20other%20debug%20test.%0A%0AScope%20the%20mock%20to%20%60%22debugpy%22%60%20only%20with%20a%20%60side_effect%60%3A%0A%0A%60%60%60python%0Aoriginal_find_spec%20%3D%20importlib.util.find_spec%0A%0Adef%20fake_find_spec%28name%3A%20str%29%20-%3E%20object%20%7C%20None%3A%0A%20%20%20%20if%20name%20%3D%3D%20%22debugpy%22%3A%0A%20%20%20%20%20%20%20%20return%20object%28%29%0A%20%20%20%20return%20original_find_spec%28name%29%0A%0Awith%20patch%28%22importlib.util.find_spec%22%2C%20side_effect%3Dfake_find_spec%29%3A%0A%20%20%20%20...%0A%60%60%60%0A%0A&repo=ibbybuilds%2Faegra">Fix All in Codex</a>

<sub>Reviews (2): Last reviewed commit: ["feat(cli): add --debug-port and --wait-f..."](https://github.com/ibbybuilds/aegra/commit/1750f4b61c6a9ea6bfd30f5976a4c242cc26a014) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=27879558)</sub>

### CODE REVIEW

File: libs/aegra-cli/src/aegra_cli/cli.py [hunk 1]
```diff
@@ -211,6 +212,19 @@ def ensure_docker_files(project_path: Path, slug: str) -> Path:
...
+    type=int,
+    help="Port for debugger to listen on (no debugger if not specified).",
+    show_default=True,
+)
+@click.option(
+    "--wait-for-client",
+    is_flag=True,
+    default=False,
+    help="Break and wait for a debugger client to attach before starting.",
+)
```
[the remark this answers is not in the document]
Comment (ibbybuilds) [reply libs/aegra-cli/src/aegra_cli/cli.py:234]: @mgrunberg can you implement this one? fail fast with `click.UsageError` when `wait_for_client` is set but `debug_port` is none.

File: libs/aegra-cli/src/aegra_cli/cli.py [hunk 2]
```diff
@@ -333,17 +353,23 @@ def dev(
...
+
+    if debug_port is not None:
+        if importlib.util.find_spec("debugpy") is None:
+            console.print(
+                "[bold red]Error:[/bold red] debugpy is not installed.\n"
+                "Install it with: [cyan]pip install debugpy[/cyan]"
+            )
+            sys.exit(1)
+        listen = f"{host}:{debug_port}" if host else str(debug_port)
+        cmd = [sys.executable, "-m", "debugpy", "--listen", listen]
```
[the remark this answers is not in the document]
Comment (ibbybuilds) [reply libs/aegra-cli/src/aegra_cli/cli.py]: @mgrunberg ok so this one matters. debugpy on a non loopback is unauthenticated arbitrary code execution in the server process, not the same risk class as uvicorn on 0.0.0.0 which has auth middleware in front of it. a user typing `--host 0.0.0.0 --debug-port 5678` is thinking "reachable from my lan" not "expose a python repl to the network."

can we do this:

1. always bind debugpy to `127.0.0.1` by default, regardless of `--host`
2. add a separate `--debug-host` flag (default `127.0.0.1`) for the rare case someone actually needs remote debugging
3. if `--debug-host` is non loopback, print a red warning panel noting the exposure

safe path stays the default, dangerous path takes a deliberate second flag. should be ~10 lines plus a doc line.

**Review by ibbybuilds — CHANGES_REQUESTED** [review]
thanks for this, core approach is solid. subprocess wrapping uvicorn with debugpy, lazy import via find_spec, graceful error path, tests cover the three main paths, docs updated in the same pr. nice work.

requesting changes on a couple small things on top of the two inline threads above:

1. version bump. both `libs/aegra-api/pyproject.toml` and `libs/aegra-cli/pyproject.toml` are still at `0.9.2`. this is a non breaking feature so patch bump both to `0.9.3`.

2. package `debugpy` as an optional extra instead of a "go install it yourself" message. in `libs/aegra-cli/pyproject.toml`:

   ```toml
   [project.optional-dependencies]
   debug = ["debugpy>=1.8"]
   ```

   and update the missing dep error to suggest `pip install 'aegra-cli[debug]'` instead of raw `pip install debugpy`. gives users a discoverable install path, keeps the version pin under our control, and doesnt bloat production installs.

once the two inline threads above and these two are addressed, this is good to go. test api was still running last i checked, make sure that goes green too.

**Comment by mgrunberg** [comment]
@ibbybuilds I did all the requested changes. After finishing the changes, I found langraph cli `--no-reload`. It could be used in combination with `debug-port`. Do you want to add it? If it out of scope, I can open another PR.

**Comment by ibbybuilds** [comment]
nice work, the `ipaddress` based parsing with ipv6 bracketing and localhost handling is cleaner than what i originally suggested. tests cover all the edge cases well.

on the `--no-reload` question: follow up pr is fine, keeps this one scoped. approving and merging now.

**Review by ibbybuilds — APPROVED** [review]
